File: examples/clj/test/com/rpl/agent/basic/rama_module_agent_test.clj
(ns com.rpl.agent.basic.rama-module-agent-test
  (:require
   [clojure.test :refer [deftest testing is]]
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.rama :as rama]
   [com.rpl.rama.test :as rtest]
   [com.rpl.agent.basic.rama-module-agent :refer [RamaModuleAgent]]))

(deftest rama-module-agent-test
  ;; Test demonstrating Rama module with agents and depot integration
  (System/gc)
  (testing "RamaModuleAgent example"
    (with-open [ipc (rtest/create-ipc)]
      (rtest/launch-module! ipc RamaModuleAgent {:tasks 1 :threads 1})

      (let [module-name (rama/get-module-name RamaModuleAgent)]
        (testing "depot is accessible and can receive data"
          (let [depot (rama/foreign-depot ipc module-name "*example-depot")]
            (is (some? depot))
            ;; Verify depot accepts data (stream processing happens asynchronously)
            (rama/foreign-append! depot "test-data")))

        (let [manager (aor/agent-manager ipc module-name)
              agent (aor/agent-client manager "FeedbackAgent")]

          (testing "agent is available in module"
            (is (contains? (set (aor/agent-names manager)) "FeedbackAgent")))

          (testing "processes feedback and returns success with message"
            (let [result (aor/agent-invoke agent "Great product!")]
              (is (= :success (:status result)))
              (is (= "Processed: Great product!" (:message result)))
              (is (= 14 (:length result)))))

          (testing "processes different feedback correctly"
            (let [result (aor/agent-invoke agent "Needs work")]
              (is (= :success (:status result)))
              (is (= "Processed: Needs work" (:message result)))
              (is (= 10 (:length result))))))))))

File: examples/java/src/test/java/com/rpl/agent/basic/RamaModuleAgentTest.java
package com.rpl.agent.basic;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentManager;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;
import java.util.HashMap;
import java.util.Set;
import org.junit.Test;

/**
 * Test class for RamaModuleAgent demonstrating Rama module with depot integration.
 *
 * <p>This test demonstrates:
 *
 * <ul>
 *   <li>Testing direct RamaModule implementation
 *   <li>Accessing and testing depot functionality
 *   <li>Verifying agent availability in the module
 *   <li>Testing agent invocations with structured responses
 * </ul>
 */
public class RamaModuleAgentTest {

  @Test
  public void testRamaModuleAgent() throws Exception {
    // Tests Rama module with manual agent topology creation
    try (InProcessCluster ipc = InProcessCluster.create()) {
      // Deploy the Rama module
      RamaModuleAgent.RamaModule module = new RamaModuleAgent.RamaModule();
      ipc.launchModule(module, new LaunchConfig(1, 1));

      // Get agent manager and client
      String moduleName = module.getModuleName();
      AgentManager manager = AgentManager.create(ipc, moduleName);

      // Verify agent is available
      Set<String> agentNames = manager.getAgentNames();
      assertTrue("FeedbackAgent should be available", agentNames.contains("FeedbackAgent"));

      AgentClient agent = manager.getAgentClient("FeedbackAgent");

      // Test first feedback processing
      HashMap<String, Object> result1 = (HashMap<String, Object>) agent.invoke("Great product!");
      assertNotNull("Agent should return a result", result1);
      assertEquals("success", result1.get("status"));
      assertEquals("Processed: Great product!", result1.get("message"));
      assertEquals(14, result1.get("length"));

      // Test second feedback processing
      HashMap<String, Object> result2 = (HashMap<String, Object>) agent.invoke("Needs work");
      assertNotNull("Agent should return a result", result2);
      assertEquals("success", result2.get("status"));
      assertEquals("Processed: Needs work", result2.get("message"));
      assertEquals(10, result2.get("length"));
    }
  }
}
